src/photon/photon_scheduler.c

```diff
@@ -23,8 +24,148 @@
 UT_icd task_ptr_icd = {sizeof(task *), NULL, NULL, NULL};
 UT_icd workers_icd = {sizeof(local_scheduler_client *), NULL, NULL, NULL};
 
+UT_icd pid_t_icd = {sizeof(pid_t), NULL, NULL, NULL};
+
 UT_icd byte_icd = {sizeof(uint8_t), NULL, NULL, NULL};
 
+void kill_process(pid_t pid) {
+  /* Kill the worker violently. */
+  kill(pid, SIGKILL);
+  waitpid(pid, NULL, 0);
+  LOG_DEBUG("Killed worker %d", pid);
+}
+
+void free_local_scheduler(local_scheduler_state *state) {
+  /* Free the command for starting new workers. */
+  if (state->config.start_worker_command != NULL) {
+    int i = 0;
+    const char *arg = state->config.start_worker_command[i];
+    while (arg != NULL) {
+      free((void *) arg);
+      ++i;
+      arg = state->config.start_worker_command[i];
+    }
+    free(state->config.start_worker_command);
+    state->config.start_worker_command = NULL;
+  }
+
+  /* Disconnect from the database. */
+  if (state->db != NULL) {
+    db_disconnect(state->db);
+    state->db = NULL;
+  }
+  /* Disconnect from plasma. */
+  plasma_disconnect(state->plasma_conn);
+  state->plasma_conn = NULL;
+
+  /* Free the list of workers and any tasks that are still in progress on those
+   * workers. */
+  for (int i = 0; i < utarray_len(state->workers); ++i) {
+    local_scheduler_client *worker =
+        *(local_scheduler_client **) utarray_eltptr(state->workers, i);
+
+    /* If the worker is a child process, we are responsible for killing it. */
+    if (worker->pid != 0) {
+      kill_process(worker->pid);
+    }
+
+    /* Free the state associated with this worker. */
+    if (worker->task_in_progress != NULL) {
+      free_task(worker->task_in_progress);
+      worker->task_in_progress = NULL;
+    }
+    free(worker);
+  }
+  utarray_free(state->workers);
+  state->workers = NULL;
+
+  /* Kill any child processes that didn't register as a worker yet. */
+  pid_t *worker_pid;
+  for (worker_pid = (pid_t *) utarray_front(state->child_pids);
+       worker_pid != NULL;
+       worker_pid = (pid_t *) utarray_next(state->child_pids, worker_pid)) {
+    kill_process(*worker_pid);
+  }
+  utarray_free(state->child_pids);
+
+  /* Free the algorithm state. */
+  free_scheduling_algorithm_state(state->algorithm_state);
+  state->algorithm_state = NULL;
+  /* Free the input buffer. */
+  utarray_free(state->input_buffer);
+  state->input_buffer = NULL;
+  /* Destroy the event loop. */
+  event_loop_destroy(state->loop);
+  state->loop = NULL;
+  /* Free the scheduler state. */
+  free(state);
+}
+
+/**
+ * Start a new worker as a child process.
+ *
+ * @param state The state of the local scheduler.
+ * @return Void.
+ */
+void start_new_worker(local_scheduler_state *state) {
+  /* We can't start a worker if we don't have the path to the worker script. */
+  CHECK(state->config.start_worker_command != NULL);
+  /* Launch the process to create the worker. */
+  pid_t pid = fork();
```

Comment:
It calls `execvp` soon after, which if successful, completely replaces the process image.

`popen` didn't seem necessary in this case, since we're already communicating with the process through sockets. Also, I needed the PID to be returned to the local scheduler so that it could later figure out if a worker process was a child or not.